### ISSUE TITLE
Request: Send last code again

### ISSUE THREAD

**Issue by JazzyPierrot** [jalvesaq/Nvim-R#408]
Hi, 
 I would really enjoy being able to send the last piece of code (whether it is a line, a chunk, a visual-selection etc.) again with one key press.

Typical use example: I send some code, get an error, go through other files to correct it, and want to test the code again. 

Sorry if it already exists but I could not find it in the documentation - and thanks for the great plugin, I enjoy writing R in vim.

**Comment by matsburg** [comment]
Here is a hacky vimscript solution. Essentially it tries to find the window with the R console buffer in, go to it and simulate pressing 'up' and 'enter' in order to re-run the last command. I think on linux the R console buffer ends in :R and on windows it starts with Rterm, so the expression `':R\|Rterm'` should hopefully match both, but may need some refinement. I noticed that this function fails when the whole .R script is sourced with e.g. `<LocalLeader>aa` but it seems to work for lines, paragraphs and visual selections.

```vim
function! LastCommand()
    let current_window_id = win_getid()
    let term_window_id = win_findbuf(bufnr(':R\|Rterm'))[0]
    call win_gotoid(term_window_id)
    let key_seq = "\<Up>\<CR>\<Esc>:call win_gotoid(" . current_window_id . ")\<CR>"
    call feedkeys(key_seq)
endfunction

nnoremap <Leader>lc :call LastCommand()<CR>

```

**Comment by jalvesaq** [comment]
What about `<LocalLeader>bb` to send marked blocks? It's not exactly what you have described but perhaps it does what you want.

**Comment by JazzyPierrot** [comment]
Thanks for your quick answers. 
I think the hack won't work with the tmux terminal, which I often use to put the terminal on another screen. However I was unaware that if I run NvimR.selection() in R after changing the file and selection (or even running single lines), I still run the first selected piece of code. This will already help. 
So there might be a solution based on :RSend NvimR.selection() and other such functions... But how to deal with single lines then ? Is there a way to store the code run by R into a variable (vim scripting novice asking) ?  

For the marked blocks, is your idea to put automatically marks each time some code is run ? Then I would still to have to go on the right buffer and put the cursor between the two marks to be able to run the code ? This would mitigate the benefits compared to being able to send exactly the same code without returning to it.

**Comment by jalvesaq** [comment]
> Is there a way to store the code run by R into a variable (vim scripting novice asking) ?

This would be very easy to implement. But I can't imagine how it could be useful. If you send the last code (either single or multiple lines) again, you will get the same result. You have to change the code to get a different result.

> For the marked blocks, is your idea to put automatically marks each time some code is run ?

No, this is your idea. What I have said was to manually put marks such as `a` and `b`. But your idea is better than mine.

Each time a selection is made, Vim automatically puts the mark `<` at first line of the selection and the mark `>` at the last line. So the solution could be a function to send the lines between these two marks regardless of how and when they were created. In fact, the function `SendSelectionToR()` already does this. So, all I have to do is:

  - Enable the key binding `<LocalLeader>ss` in normal mode too (it is currently enabled only in Visual mode).

  - If in Normal mode, check if the cursor is between the marks `<` and `>` before sending the last selection of lines.

**Comment by jalvesaq** [comment]
Please, try the branch [resend_selected_lines](https://github.com/jalvesaq/Nvim-R/tree/resend_selected_lines).

**Comment by matsburg** [comment]
Here is another solution, the workflow is:
1. Select some code in visual mode and hit `<Leader>sc` to both send the selection to R and "store" it.
2. In normal mode, hit `<Leader>sc` to run the "stored" code again. This should work from any buffer/window/tab.

This is achieved by setting global marks A and B when the code is first selected. These global marks can then be jumped to from any buffer.

```vim
function! StoreCommand() range
    call SendSelectionToR("echo", "stay")
    normal! `<mA`>mBgv
endfunction

function! RunStoredCommand()
    normal! mC`Av`B
    call SendSelectionToR("echo", "stay")
    execute "normal! \<Esc>`C"
    " For some reason if you call this function from a different tab
    " then it puts you in insert mode after jumping back to C, hence
    " the need for stopinsert
    stopinsert
endfunction

vnoremap <Leader>sc :call StoreCommand()<CR>
nnoremap <Leader>sc :call RunStoredCommand()<CR>
```

P.S. @jalvesaq I like the change you've made in the new branch, and I think I would personally find it useful in general, beyond just this issue.

**Comment by JazzyPierrot** [comment]
Thanks again to both of you. 
`If you send the last code (either single or multiple lines) again, you will get the same result`
Yes, right, I did not think this through, but my idea was to keep track of lines, selected code or chunk sent, and to ignore sourced files and packages. But marking specifically one piece of code to be rerun in the future is probably a better practice. 

The addition to be able to run a selection in normal mode can avoid a couple of keystrokes (I just discovered the **gv** command, thanks !) and I don't see any downside, so it is a nice addition. 

@matsburg This works well so far, thanks, I just removed the C mark with 
```
 normal! `Av`B
 execute "normal! \<Esc>`2\<C-O>"
```
I am going to seek the best workflow with these elements, thank you for your help. I think I can close this.
